[MRG] hash checking --update option for upload, fetch, and clone

benlindsay opened the pull request:
I started working on adding a hash checking option, and this seems to work so far. I haven't added tests or anything yet, and this relies on the change in #142 which is unmerged, and which I copied into these changes. I just wanted to get input on this before putting too much time into it.

My thought, based on @betatim's suggestion in #138 of adding an `--update` flag, is that, assuming a file already exists on OSF, if `--force` is specified, we don't care whether `--update` is specified, and upload without comparing the hashes. If `--force` is not specified, then we check the hash if `--update` is specified, and fail the upload otherwise. Does that seem like a reasonable way to do it?

Also note that I copied the local checksum function from https://gist.github.com/rji/b38c7238128edf53a181#file-sha256-py. It seems to work but I don't know if there's a better way to do it.

Also, once we settle on the right way to do this, I figure we could do the same thing for `fetch` as well.

# EDIT

Forgot to add Fixes #138 

Also I think this Fixes #15

benlindsay commented:
I added the update option to fetch and clone as well since that was a pretty easy thing to do. Let me know what you think of these options. For fetch, `osf fetch -u path/to/file.txt` will check the hashes, update if different, or report that the files are the same otherwise. For clone, `osf clone -u` will skip files with matching hashes with no reporting other than the progress bar saying how many files have been downloaded. If this sounds reasonable, I can start thinking about tests.

betatim commented:
Maybe use `-U` for update?

betatim commented on osfclient/cli.py:
```diff
@@ -200,6 +204,10 @@ def fetch(args):
+            if local_path_exists and not args.force and args.update:
+                if file_.hashes['sha256'] == sha256_checksum(local_path):
+                    print("Local file %s already matches remote." % local_path)
+                    break
```
Should this be `continue` instead of `break`? From reading the local code it seems we are iterating over all the files in a project and if one of them doesn't need updating we should skip it but continue with the rest. Instead of ending the `for` loop which is what `break` does. If we should use a `break` then I think we should have a comment explaining why, because I can already see future Tim coming back here and being confused again.

betatim commented on osfclient/models/storage.py:
```diff
@@ -48,7 +49,7 @@ def files(self):
                                    self._files_key)
 
-    def create_file(self, path, fp, update=False):
+    def create_file(self, path, fp, force=False, update=False):
```
Can you extend the doc string and point out how force and update work together/are different from each other.

benlindsay replied:
Right, forgot to update the doc string, I'll get right on that.

betatim commented on osfclient/models/storage.py:
```diff
@@ -82,13 +83,18 @@ def create_file(self, path, fp, update=False):
                 for file_ in self.files:
                     if norm_remote_path(file_.path) == path:
+                        if not force:
+                            if sha256_checksum(path) == file_.hashes['sha256']:
```
Does it happen that `file_.hashes` doesn't contain the key `sha256`? If it can happen we should use `file_.hashes.get('sha256')` everywhere.

betatim commented:
Looks pretty good to me, just a few comments and questions.

benlindsay commented:
Thanks! I'll work on those changes. Also could you look at my [checksum implementation](https://github.com/osfclient/osfclient/pull/146/commits/67c23901917cd3bf55ef93481be1840999cab465#diff-5c2f806a96651664efc7a0f319a17fd4R68)? I copied it from a github gist, and I don't know why it has a default block size of 65536 (not sure what the significance of 2^16 is), and I'm not sure if it's good or bad practice to put a link to the source for little code snippets like that.

betatim commented:
I was wondering about the checksum implementation but assumed that it was "correct". Meaning it gives the same checksum as OSF for the same file. I am not sure what the significance of the precise value of the blocksize is. You need to hash your data in blocks, so you need to choose a size that isn't too big but also not too small :-/

re: link to source: for sure put a link to where the code came from. Ideally we'd find a place to copy it from where the license is clear.

benlindsay commented:
Looking at comments under [this SO answer](https://stackoverflow.com/a/44873382/2680824), somewhere between 4 kb and 128 kb seem to be common choices, so 64 kb kind of falls in the middle of those, so hopefully that's fine. Whether I use the md5 or sha256 hash, the results do match up with the hashes provided by OSF.

I modified the checksum function, and switched to md5 since it's supposed to be a little faster than sha256, and left in a link to the above SO post that was my main source of info.

I'll look into adding tests for this soon, but I'm new to the pytest framework so might need some time.

benlindsay commented:
@betatim I was able to update the existing tests to accommodate the `update` option, but adding a new test will require a little more understanding of this testing setup. I'm thinking the way to test the `update` option would be to hardcode the `md5` hash attribute for a mock file object, and check to make sure an API call is made if the hashes differ. Is that how you would do it, and either way if you could give me some quick pointers on how to set it up I'd appreciate it.

benlindsay commented:
Never mind, I think I got it well enough figured out, although this is my first dive into pytest, so there are likely some things I could have done better, so please let me know if you see any. All the tests pass, and I think they cover the major cases that the `--update` option brings up, so I think it's about ready to merge after review.

benlindsay commented:
== EDIT ==

For peace of mind, I set up Travis on my fork. (https://travis-ci.org/benlindsay/osfclient) All tests passed there too.

[![Build Status](https://travis-ci.org/benlindsay/osfclient.svg?branch=triggertravis)](https://travis-ci.org/benlindsay/osfclient)

benlindsay commented:
Side question: I'm seeing upload speeds of no less than 16 seconds per file, sometimes about 30 seconds, even for very small or empty files. I haven't gone through rigorous tests, but with some quick tests, I'm not seeing any time difference between uploading a new file vs uploading the file with --update. Any idea why that might be?

benlindsay commented:
Shoot, I just realized this whole pull request is me falling victim to the root of all evil -- premature optimization. I did some quick time profiling and found that for a project I'm working on that has a fair number of files, by far the biggest time waster is looping over the files to find the right upload URL ([link to line](https://github.com/osfclient/osfclient/blob/03be3008368cda2557fe9fabfbfa96c14da580bd/osfclient/models/storage.py#L90)). The actual upload is pretty fast in comparison, although I have only tested this on files under 100 MB.

I think this whole pull request is kind of useless without speeding up the process of finding the right upload URL. I'm not sure if it's feasible to speed that up, but I'll create an issue for it to discuss it.

betatim commented:
In principle yes lets merge this. Though travis hasn't run on this branch which makes me suspicious we broke it somewhere :-/

betatim commented:
> if the offer is still on the table, now that I'm a little more familiar with the code, I'd be interested in accepting merge rights if that will reduce your load a little.

Of course! You should have an invite and welcome aboard. Remember with great power comes great responsibility ;) About the only rule is that no one can merge their own PRs.

---

Let's work out #151 (and failures that have crept in?) then return to this PR.

betatim commented:
Closing and opening this PR to trigger travis.

betatim commented:
Huzzah!

colinbrislawn commented:
I'm so glad to see this functionally being added to  osf client! Is this ready to go live?

benlindsay commented:
@colinbrislawn Yes, I think we're just fixing the build error due to problems with the pypi test upload first. (#154)

benlindsay commented:
@betatim Now that the Travis is happy with the test PyPI uploads, how do you feel about merging this one? I don't know how you typically decide when to bump the version number, but it would be nice to see this and the large file connection error fix pushed up to production PyPI soon.

benlindsay commented:
@betatim Now that most of the administration stuff is out of the way, I just want to bump this one up. Can we merge this or is there something else I should add/change first?

benlindsay commented:
@betatim I know you're busy, so I just want to give this another friendly bump. I'm trying to stick with the "don't merge your own pull request" rule, but that big green "Merge pull request" button sure is tempting... 😉

ctb commented:
nice!
